### PR TITLE
Improve Interactive Layer Transforms Usability

### PR DESCRIPTION
**Pull request by Tobias314**
### Summary

Several usability improvements for the interactive layer transforms editor.

**Layer center as pivot.** The 7-matrix live transform chain scaled and rotated around the *dataset* center, so rotating a layer that does not sit in the middle of the dataset swung it around a far-away point instead of turning it in place. The pivot is now the center of the layer's own bounding box, and `buildLiveTransforms` takes that pivot directly instead of a dataset bounding box.

Since we store the pivot point in the transforms, transforms that were stored with the old pivot are rebased onto the new one. Without this, any transforms created before the change would still pivot around the dataset center. With the rebasing (writing the layer center as pivot and changing the translation transform in a way that the overall transform chain stays identical) all layers now rotate around the layer center. The only visible change is that the translation value shown in the popup for dataset that previously had a different pivot is now changed since it gets adapted to the layer center pivot.

**Relative translation sliders.** The translation rows used a fixed step of 1 voxel and a range of ±the dataset extent. Zoomed out, a 1-voxel step is far below one screen pixel and dragging does nothing visible; zoomed in, the range makes the slider hopelessly coarse.

The sliders are now *relative*: the handle rests in the center and does not show the value at all — the number input next to it does. Dragging right increases the translation, dragging left decreases it, live. On release the handle returns to the center, so the next gesture continues from the value just reached and the translation itself is no longer limited by the slider's range. The reachable increment in one gesture is one viewport extent in either direction.

**Relative scaling sliders.** The scaling rows work the same way, but multiplicatively and in log scale: the rail spans one decade in either direction, i.e. from a tenth of the current size up to ten times it, labeled `×1/10` and `×10`. Equal distances on the rail are equal factors, so shrinking and enlarging a layer are symmetric gestures.

**Shared component.** Both live in one `RelativeSlider` (`viewer/view/left_border_tabs/components/relative_slider.tsx`). The gesture logic — capture the base value when the gesture starts, apply live, recenter on release, keep the rail stable while dragging — is in one place; what differs is a `RelativeSliderConfig` describing the range, step, labels and how a handle position turns into a new value (adding voxels vs. multiplying). `AxisSliderRow` gained an optional `sliderNode` so the row layout (label, number input, flip and reset buttons) stays shared with the rotation rows, which keep their absolute slider.

### Steps to test:
- Open a dataset and edit the transform via the sliders, the dataset should pivot (rotate/scale) around the layer center
- Any transformation (translation/scale/rotation) should be reachable (no maximum values)
- Translation slider windows should adapt to viewport size

**Edge cases**
- Open the popover for a layer with no transforms at all: sliders start centered, translation and rotation at 0 and scaling at 1.
- Open the popover for a layer with an old (non-layer-center) pivot -> should still pivot around layer center
- Open a layer whose `coordinateTransformations` are not the editable 7-matrix pattern (e.g. a thin-plate-spline, or a chain with an asymmetric pivot pair written via `api.data.setLayerTransforms`): the popover must render the "not editable here" message rather than crashing or offering sliders.
- Open the popover in a view mode without ortho viewports (e.g. one viewport maximized): the translation sliders must still be draggable (they fall back to ±100 voxels).

------
(Please delete unneeded items, merge only when none are left open)
- [x] Added changelog entry (create a `$PR_NUMBER.md` file in `unreleased_changes` or use `./tools/create-changelog-entry.py`)
- [ ] Updated [documentation](../blob/master/docs) if applicable (should we do this here?)
- [x] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)

### CODE REVIEW

File: frontend/javascripts/test/admin/compose_voxel_size.spec.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    const existing = { type: "affine" as const, matrix: [[1, 0, 0, 5]] as any };
+    const layersWithExisting = [
+      { ...makeLayerLink("coarse", "coarse_layer"), transformations: [existing] },
+    ];
+
+    const [layer] = withVoxelSizeTransforms(layersWithExisting, datasets, targetVoxelSize);
+
+    expect(layer.transformations).toHaveLength(1 + EXPECTED_LIVE_TRANSFORMATION_LENGTH);
+    expect(layer.transformations[0]).toBe(existing);
+  });
```
Comment (Tobias314) [inline frontend/javascripts/test/admin/compose_voxel_size.spec.ts]: Basically, we keep existing transforms, and put the scaling on top. However, this will lead to layer transforms that are not editable via the UI. Not sure, whether this is what we want. Alternatively, we could think about forbidding this case.

File: frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        // There is no meaningful filled part of the rail for a handle that is always centered. The
+        // dots that the marks add to the rail's ends stay, since they tie the labels to the rail.
+        included={false}
+        tooltip={{ formatter: (v) => (v == null ? "" : activeConfig.formatOffset(v)) }}
+        ariaLabelForHandle={ariaLabel}
+        onChange={handleChange}
+        // The value has already been applied by the last onChange, so this only recenters the handle.
+        onChangeComplete={endAction}
+        style={{ marginBottom: MARK_LABEL_MARGIN }}
+      />
```
[the remark this answers is not in the document]
Comment (Tobias314) [reply frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx]: It exists and is also part of antd's type declaration but it seems to be not in the docs. Let's keep it for accessibility reasons. The only danger is a change in future antd versions and this should be easy to spot by the type checker.

**Comment by MichaelBuessemeyer** [comment]
Thanks for the PR Tobi! And sorry to bump in: I just noticed

> Layer center as pivot. The 7-matrix live transform chain scaled and rotated around the dataset center, so rotating a layer that does not sit in the middle of the dataset swung it around a far-away point instead of turning it in place. The pivot is now the center of the layer's own bounding box, and buildLiveTransforms takes that pivot directly instead of a dataset bounding box.

I initially designed it explicitly this way as this was at least at must for the ortho rotation feature (e.g. alt + arrow keys). When I first read this, I am not 100% sure whether that might not break some things / lead to unexpected behaviour. At least I just recall having problems when it was not the center for the ortho rotation feature. => This might apply here. So what I am saying: Did you check this? Sorry to go sure on this maybe this just isn't a problem at all and I am jumping to conclusions. On a second longer thought while writing this, I also can't really come up with a problem. 

The only scenario is in case one has layer where the centers do not equal but all of them should be rotated by e.g. 90° and then still look like everything was rotated by 90°. This case would then no longer be supported by the transforms if the rotation center is the layer bbox center and not the ds bbox center. Or at least getting this correct would be very hard.


So yeah, I just wanted to mention my thoughts when I read `Layer center as pivot.` which screwed over the whole ortho rotation feature during development at some stage ✌️

**Comment by philippotto** [comment]
> The only scenario is in case one has layer where the centers do not equal but all of them should be rotated by e.g. 90° and then still look like everything was rotated by 90°. This case would then no longer be supported by the transforms if the rotation center is the layer bbox center and not the ds bbox center. Or at least getting this correct would be very hard.

You buried the lead a bit here, but yes, that's the actual problem. There are two scenarios for rotation:
- Scenario 1: all layers in a dataset already fit together (but are sized differently) and a user wants to rotate the entire dataset (pivot = dataset center)
- Scenario 2: layers in a dataset don't match yet and the user wants to align these by rotating individual layers (pivot = layer center)

We should think through how to support both scenarios. I'm not that familiar with the whole transformation definition code. @MichaelBuessemeyer Do you think this is easy to accomplish? Scenario 1 only happens in the dataset settings. Scenario 2 only happens in the live transforms UI. Can we simply use different rotation-builder-functions in these two places?

**Comment by Tobias314** [comment]
> * Scenario 1: all layers in a dataset already fit together (but are sized differently) and a user wants to rotate the entire dataset (pivot = dataset center)
> * Scenario 2: layers in a dataset don't match yet and the user wants to align these by rotating individual layers (pivot = layer center)
> 
> We should think through how to support both scenarios.

Yes, looks like we need both. I have mostly focused on scenario 2 since that was what I needed and implemented for this PR and my project. We should probably test whether scenario 1 still works with those changes. I haven't done this so far.

Conceptionally, this should not bee too hard. You can express any of our SRT transformations around a pivot q as an equivalent SRT' transformation around a new pivot p (so basically, you only have to adapt the translation part to get an equivalent transform around a different pivot). This "rebasing" is already implemented in the PR to keep backwards compatibility. So I think, for each scenario, we can just first rebase any existing SRT transform to the pivot we need and then do the transformation. This PR already does this for the interactive layer transforms. Maybe, we can add this to the ortho rotation as well? @MichaelBuessemeyer

**Comment by normanrz** [comment]
I think in 95% of cases, people want to uniformly scale the layers. A locking mechanism that is on by-default would be great!

**Comment by MichaelBuessemeyer** [comment]
> This PR already does this for the interactive layer transforms. Maybe, we can add this to the ortho rotation as well? @MichaelBuessemeyer

You mean the feature that rotates around the 3 axis arbitrarily via shortcuts? e.g. when the center is at `100,100,100` but you are viewing at `10,10,10` you want the rotation center to be at `10,10,10` as well? That's how it already works 🤔.

> I think in 95% of cases, people want to uniformly scale the layers. A locking mechanism that is on by-default would be great!

Oh yeah, that sounds like good input 👍

**Comment by Tobias314** [comment]
> You mean the feature that rotates around the 3 axis arbitrarily via shortcuts? e.g. when the center is at `100,100,100` but you are viewing at `10,10,10` you want the rotation center to be at `10,10,10` as well? That's how it already works 🤔.

Not really. As I understand it, the ortho rotation feature does a rotation of all layers around the dataset center, right? So we need the dataset center as pivot for this. However, if someone used the layer transformation UI (that this PR implements) before doing the ortho rotation, the transforms for any transformed layer will have the layer center set as pivot.
So we need to make sure that we first "rebase" the layer transforms from using the layer center as pivot to using the dataset center as pivot (can be done with the rebasing function implemented in this PR) before applying any the ortho rotation. @MichaelBuessemeyer do you agree?

**Comment by MichaelBuessemeyer** [comment]
> So we need to make sure that we first "rebase" the layer transforms from using the layer center as pivot to using the dataset center as pivot (can be done with the rebasing function implemented in this PR) before applying any the ortho rotation. @MichaelBuessemeyer do you agree?

No I don't think we need to consider the ortho rotation feature here at all. It is completely independent from the layer transforms and works differently. The layer transforms reposition / transform the layers in the dataset space. The ortho rotation feature rotates all viewports along the three rotation axis at the current position in the 3d space to be able to look at the dataset from different angles and the standard 90°. So no rebase needed or so. When working on this feature, this should never be influenced / influence the ortho rotation feature. One transforms layer the other is just useful to view the data from different angles. These are two different things. But one can theoretically transform a layer in such a way to get the same perspective as when untransformed and then rotating the viewports but still, they should be considered independent implementation wise.

**Comment by Tobias314** [comment]
> No I don't think we need to consider the ortho rotation feature here at all.

As discussed offline, I think there was some confusion due to wrong terminology. When I said "ortho rotation feature", I actually meant the feature where you can rotate the full dataset in 90° steps in the dataset settings.

After checking this, we found that this 90° step rotation feature is disabled if we ever edited the layer transforms via the UI. So we should never run into a problem with wrong pivots there and everything should already work as intended. :tada:

**Comment by philippotto** [comment]
Great! I'll review today then. Testing already went well :tada:

File: frontend/javascripts/viewer/model/accessors/dataset_layer_transformation_accessor.ts [hunk 3]
```diff
@@ -721,21 +733,65 @@ export function extractSRTFromTransforms(transforms: CoordinateTransformation[])
...
 
+// Extracts the pivot (the point that scaling and rotation happen around) from a 7 matrix
+// coordinate transformation of a layer. It is stored as the negated translation of the first matrix.
+export function extractPivotFromTransforms(transforms: CoordinateTransformation[]): Vector3 | null {
+  if (transforms.length !== EXPECTED_LIVE_TRANSFORMATION_LENGTH) return null;
+  const [x, y, z] = extractTranslationFromMatrix(transforms[0] as AffineTransformation);
+  return [-x, -y, -z];
+}
+
+// Returns the translation that expresses the same overall transform around a different pivot.
```
Comment (philippotto) [inline frontend/javascripts/viewer/model/accessors/dataset_layer_transformation_accessor.ts:744]: I find this comment confusing. The function returns a translation (vector3) and consumes an input transform (`SRTValues`). How can the Vector3 "express the same overall transform" as the input? From reading the tests, I think I understood it now, but the comments here should be a lot clearer.

**Review by philippotto — COMMENTED** [review]
I haven't reviewed everything yet but wanted to leave some initial feedback. I really like the new relative sliders :+1: they feel nice. however, it would make sense to explain their behavior somewhere in the UI. two ideas:
- in the transforms dialog below the headline
- or: in a tooltip on the sliders

Something like "The relative sliders snap back to the center to allow for fine as well as large changes."

<sub>Commented in [CodeRabbit Change Stack](https://app.coderabbit.ai/change-stack/scalableminds/webknossos/pull/9899)</sub>

**Comment by MichaelBuessemeyer** [comment]
I just unsubscribed from the PR to get less notifications. If I should join back in, please notify me somehow 😇

File: frontend/javascripts/test/model/accessors/dataset_layer_transformation_accessor.spec.ts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    const srt: SRTValues = { scale: [2, 3, 4], rotation: [10, 20, 30], translation: [5, 6, 7] };
+
+    expect(rebaseTranslationToPivot(srt, PIVOT, PIVOT)).toEqual([5, 6, 7]);
+  });
+
+  it("should leave the translation untouched for a pure translation", () => {
+    // With an identity linear part the pivot is irrelevant.
+    const srt: SRTValues = { scale: [1, 1, 1], rotation: [0, 0, 0], translation: [5, 6, 7] };
+
+    almostEqual(expect, rebaseTranslationToPivot(srt, [1, 2, 3], [400, -5, 60]), [5, 6, 7], 1e-9);
```
Comment (philippotto) [inline frontend/javascripts/test/model/accessors/dataset_layer_transformation_accessor.spec.ts:182]: how about putting 1e-9 in a constant? also would have the potential to better explain the difference to the already existing epsilon variable.

File: frontend/javascripts/test/model/accessors/dataset_layer_transformation_accessor.spec.ts [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+  });
+
+  it("should yield (I - A) * p for the classic offset case", () => {
+    // t = 0 and the new pivot at the origin, with a pure scaling: t' = p - A p = p * (1 - scale).
```
Comment (philippotto) [inline frontend/javascripts/test/model/accessors/dataset_layer_transformation_accessor.spec.ts:186]: what are I and A? should be explained here.

File: frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+// pointer can resolve and never quantizes a drag.
+export const TRANSLATION_SLIDER_STEP = 1;
+
+// The smallest scaling that can be reached, by the scale slider as well as by the scale number input.
+export const MIN_SCALE = 0.0001;
+
+// The scaling slider works in log10 space, so that shrinking and enlarging a layer are symmetric
+// gestures: one decade in either direction, i.e. a tenth of the current size up to ten times it.
+const SCALE_SLIDER_RANGE = 1;
+// About 2.3% per step, fine enough that a drag is not visibly quantized.
```
Comment (philippotto) [inline frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx:20]: where is 2.3 coming from?

File: frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+const SCALE_SLIDER_STEP = 0.01;
+
+// Ranges from this value on are labeled with an exponent (2e4 instead of 4000), so that the labels stay short.
```
Comment (philippotto) [inline frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx:23]: 2e4 is not 4000

File: frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+// Applies a slider offset to the value the current translation slider action started from.
+// Rounding keeps the float dust of the summation out of the number input.
+export function applyRelativeTranslationDelta(base: number, delta: number): number {
+  if (!Number.isFinite(base)) {
+    return Number.isFinite(delta) ? delta : 0;
+  }
+  if (!Number.isFinite(delta)) {
+    return base;
+  }
+  return Math.round((base + delta) * 1e4) / 1e4;
```
Comment (philippotto) [inline frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx:105]: why not simply `base + delta`? arent both integers?

File: frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+  const exponent = Math.floor(Math.log10(value));
+  const mantissa = Math.round(value / 10 ** exponent);
+  // Rounding up from 9.5 leaves the single-digit range, as can an imprecise logarithm.
+  return mantissa >= 10 ? { mantissa: 1, exponent: exponent + 1 } : { mantissa, exponent };
+}
+
+// Turns a viewport extent into a usable slider range, rounded to a single significant digit (e.g.
+// 1253 becomes 1000) so that the largest increase/decrease one slider action can apply is a round
+// number (round numbers are easier to display as slider labels).
+export function translationSliderRangeFromViewportExtent(extentInVoxel: number): number {
```
Comment (philippotto) [inline frontend/javascripts/viewer/view/left_border_tabs/components/relative_slider.tsx:61]: for me, a range is usually a two-number value. this function returns a single number. if I understood it correctly, that value will be used as like `range = [-value, +value]`. so, the "length" of the actual range will be twice as big as the returned value here. my suggestion would be to either rename the function (e.g., a tongue twister like "halfRangeLength") OR add a comment that this will be only half of the actually covered length OR just change the function to return `[value, value]`. I prefer the last option.

File: frontend/javascripts/test/viewer/relative_slider.spec.ts [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+describe("applyRelativeFactor", () => {
+  it("should scale relative to the base value", () => {
+    expect(applyRelativeFactor(4, 0)).toBe(4);
+    expect(applyRelativeFactor(4, 1)).toBe(40);
+    expect(applyRelativeFactor(0.5, -1)).toBe(0.05);
+  });
+
+  it("should keep four significant digits, so that tiny scalings stay usable", () => {
+    expect(applyRelativeFactor(0.001, 0.1)).toBeCloseTo(0.001259, 6);
```
Comment (philippotto) [inline frontend/javascripts/test/viewer/relative_slider.spec.ts:139]: please add a comment à la: `// 0.001 * 10**0.1 ~ 0.001259`

File: frontend/javascripts/viewer/view/left_border_tabs/components/layer_transform_settings_popover.tsx [hunk 11]
```diff
@@ -135,15 +192,7 @@ export function LayerTransformSettingsContent({
...
-  const datasetBbox = getUntransformedDatasetBoundingBox(dataset);
-  const translationSettingLimits = useMemo<[number, number, number]>(
-    () => [
-      datasetBbox.max[0] - datasetBbox.min[0],
-      datasetBbox.max[1] - datasetBbox.min[1],
-      datasetBbox.max[2] - datasetBbox.min[2],
-    ],
-    [datasetBbox],
-  );
+  const datasetBbox = useMemo(() => getUntransformedDatasetBoundingBox(dataset), [dataset]);
```
Comment (philippotto) [inline frontend/javascripts/viewer/view/left_border_tabs/components/layer_transform_settings_popover.tsx:195]: the function is already memoized.

```suggestion
const datasetBbox = getUntransformedDatasetBoundingBox(dataset);
```

File: frontend/javascripts/viewer/view/left_border_tabs/components/layer_transform_settings_popover.tsx [hunk 12]
```diff
@@ -255,6 +342,12 @@ export function LayerTransformSettingsContent({
     handleChange({ scale: newScale, rotation, translation });
   };
 
+  // The scaling row shows and edits only the magnitude; the flip orientation (the sign of the scale)
+  // is kept as it is, since the flip toggle lives in the rotation row.
+  const updateScaleMagnitude = (axis: 0 | 1 | 2, magnitude: number) => {
+    updateScale(axis, magnitude * (scale[axis] < 0 ? -1 : 1));
```
Comment (philippotto) [inline frontend/javascripts/viewer/view/left_border_tabs/components/layer_transform_settings_popover.tsx:348]:
```suggestion
    updateScale(axis, magnitude * Math.sign(scale[axis]));
```

File: frontend/javascripts/viewer/view/left_border_tabs/components/layer_transform_settings_popover.tsx [hunk 13]
```diff
@@ -319,12 +420,18 @@ export function LayerTransformSettingsContent({
...
-          min={0.0001}
-          max={10}
-          step={0.1}
-          // The slider shows only the magnitude; keep the current flip orientation here. Resetting
-          // the flip is handled by the rotation row, where the flip toggle lives.
-          onChange={(v) => updateScale(i as 0 | 1 | 2, v * (scale[i] < 0 ? -1 : 1))}
+          inputMin={MIN_SCALE}
+          step={SCALE_INPUT_STEP}
+          onChange={(v) => updateScaleMagnitude(i as 0 | 1 | 2, v)}
+          onCommit={(v) => updateScaleMagnitude(i as 0 | 1 | 2, Math.max(MIN_SCALE, Math.abs(v)))}
```
Comment (philippotto) [inline frontend/javascripts/viewer/view/left_border_tabs/components/layer_transform_settings_popover.tsx:426]: when is this onCommit necessary? when I drag the scale, I cannot get the value below the absolute minimum of 0.0001 (even while dragging without releasing the mouse). maybe the whole onCommit code can be removed? also, I think it's good that the values don't change on mouse-up.

File: unreleased_changes/9899.md [hunk 14]
```diff
@@ -0,0 +1,4 @@
+### Changed
+- Layer transforms are now scaled and rotated around the center of the layer instead of the center of the dataset, so that a layer rotates in place. Existing transforms are automatically rebased onto the new pivot, which does not change how the layer is rendered.
```
Comment (philippotto) [inline unreleased_changes/9899.md:2]: please clarify that this only refers to the "live transform UI". the dataset rotation feature in the dataset settings are not affected.

File: unreleased_changes/9899.md [hunk 15]
```diff
@@ -0,0 +1,4 @@
...
+- The translation sliders of the layer transform settings are now relative: their handle rests in the center and dragging it increases or decreases the translation by up to one viewport extent, labeled at both ends of the slider. On release the handle returns to the center, so that the translation is not limited by the slider's range and the reachable precision follows the zoom level.
+- The scaling sliders of the layer transform settings are relative as well: they work in log scale and reach from a tenth of the current size up to ten times it, so shrinking and enlarging a layer are symmetric gestures.
```
Comment (philippotto) [inline unreleased_changes/9899.md:4]: I think the log scale is not that important to the users. "they reach from a tenth..." should be enough in my opinion and is more concise.

**Review by philippotto — COMMENTED** [review]
<sub>Commented in [CodeRabbit Change Stack](https://app.coderabbit.ai/change-stack/scalableminds/webknossos/pull/9899)</sub>

**Review by philippotto — CHANGES_REQUESTED** [review]
cool stuff! see my comments and also the ones of code rabbit ("mark as resolved" when you think it makes sense).
